design/runtime-parameters.md

```diff
@@ -0,0 +1,339 @@
+# Runtime and type-safe parameters
+
+We need a way to take run-time inputs, giving pipeline authors the right amount of control / expressivity.
+YAML pipelines already accept parameters when used as part of a template.
+Runtime parameters are a natural evolution of that syntax.
+Along the way, we will augment the capabilities of template parameters in a natural way.
+
+Scenarios:
+- Pipeline users can supply different values to tools and tasks at run time
+- Pipeline authors can control the types, ranges allowed, and defaults for run-time parameters
+- Pipeline authors can allow flexible, run-time control over stages/jobs to run, including adding/removing matrix legs
+- The pipeline itself can be altered at run time: pools, agentspecs, service connections, environments, and more
+- The runtime values can be used in template expressions (`${{ if }}`, for example), for instance so that jobs/steps can be dynamically selected (this is not currently possible because queue-time variables are not available during template expansion)
+
+## Syntax
+
+```yaml
+parameters:
+# this is a sequence, so the block may be repeated as many times as needed
+- name: string          # name of the parameter; required
+  displayName: string   # UI string
+  type: enum            # data type; see below
+  default: any          # default value; if no default, then the parameter MUST be given by the user at runtime
+  values: [ string ]    # allowed list of values (for some data types)
+  secret: bool          # whether to treat this value as a secret; defaults to false
+```
+
+See [below](#back-compat-with-current-syntax) for details about how this coexists with existing `parameters` syntax.
+
+## Run panel ("queue dialog")
+
+Given this YAML:
+
+```yaml
+parameters:
+- name: myString
+  type: string
+  default: 'default value!'
+- name: mynum
+  displayName: 'My Number Param'
+  type: number
+  default: 123
+- name: mypool
+  displayName: 'Pool'
+  type: pool
+  values:
+  - ubuntu-latest
+  - windows-latest
+  - macos-latest
+
```

Comment:
In classic ADO build or release we go all with Variables and Variable groups. Here in YAML we have something like parameters and also variables. During build save or queue time does it show up parameters as variables or variables itself? What about global variables and system defined variables to display at queue time for user to change it if needed. Ex: System.Debug = true or false.